Subject: Save the current patchset in the kernel
From: Carsten Emde <C.Emde@osadl.org>
Date: Wed,  9 Jan 2013 08:44:49 +0100

In the good old days when only a small number of computer boards were
released per year and mainline kernel support was available for most of
them, it was an excellent idea to save the current kernel configuration
in the binary kernel. This made it possible to reproduce a kernel from
the vanilla kernel sources at any time.

In modern times, however, when myriads of (mostly ARM based) computer
boards are inundating the market every year most of which require a
large patchset that often is difficult to obtain, the saved config file
is nearly useless. (If a patchset cannot be obtained with reasonable
effort, the board vendor, of course, is in breach of the Linux kernel
license.)

This patch now provides a mechanism to store the current patchset in the
binary kernel in a very similar way as it is done with the config file. To
reproduce the source tree that was used for a particular kernel, use the
command sequence:
  tar zxf /proc/patchset.tar.gz baseversion
  major=`cut -d. -f1 baseversion`
  urldir=http://www.kernel.org/pub/linux/kernel/v$major.x
  dir=linux-`cat baseversion`
  rm -f baseversion
  archive=$dir.tar.xz
  wget $urldir/$archive
  tar Jxf $archive
  cd $dir
  tar zxf /proc/patchset.tar.gz
  quilt push -a
  zcat /proc/config.gz >.config

BTW: Making the kernel sources available through this mechanism does not
entirely fulfill the disclosure obligations of the GPL, but it certainly
is much better than making available nothing.

Signed-off-by: Carsten Emde <C.Emde@osadl.org>

---
 init/Kconfig               |   45 ++++++++++++++++++++
 kernel/Makefile            |   17 +++++++
 kernel/patchset.c          |  100 +++++++++++++++++++++++++++++++++++++++++++++
 scripts/Makefile           |    1 
 scripts/bin2c.c            |   36 ++++++++++++++++
 scripts/extract-ikpatchset |   68 ++++++++++++++++++++++++++++++
 6 files changed, 267 insertions(+)

Index: linux-6.18.2/kernel/Makefile
===================================================================
--- linux-6.18.2.orig/kernel/Makefile
+++ linux-6.18.2/kernel/Makefile
@@ -91,6 +91,7 @@ obj-$(CONFIG_UTS_NS) += utsname.o
 obj-$(CONFIG_USER_NS) += user_namespace.o
 obj-$(CONFIG_PID_NS) += pid_namespace.o
 obj-$(CONFIG_IKCONFIG) += configs.o
+obj-$(CONFIG_IKPATCHSET) += patchset.o
 obj-$(CONFIG_IKHEADERS) += kheaders.o
 obj-$(CONFIG_SMP) += stop_machine.o
 obj-$(CONFIG_AUDIT) += audit.o auditfilter.o
@@ -163,6 +164,22 @@ filechk_cat = cat $<
 $(obj)/config_data: $(KCONFIG_CONFIG) FORCE
 	$(call filechk,cat)
 
+$(obj)/patchset.o: $(obj)/patchset_data.h
+
+# patchset_data.h contains the patchset as gzipped tar archive
+# The archive can be extracted from /proc/patchset.tar.gz
+baseversion: include/generated/utsrelease.h
+	@cut -d" " -f 3 include/generated/utsrelease.h | tr -d '"' | cut -d- -f1 >$@
+targets += patchset_data.gz
+$(obj)/patchset_data.gz: $(wildcard patches/*) patches/series baseversion
+	@tar zcf $@ `quilt applied | sed '/^patches\//! { s,\(.*\),patches/\1, }'` patches/series baseversion
+
+filechk_ikpatchsetgz = (echo "static const char kernel_patchset_data[] __used = MAGIC_START"; cat $< | scripts/bin2c; echo "MAGIC_END;")
+
+targets += patchset_data.h
+$(obj)/patchset_data.h: $(obj)/patchset_data.gz FORCE
+	$(call filechk,ikpatchsetgz)
+
 # kheaders_data.tar.xz
 $(obj)/kheaders.o: $(obj)/kheaders_data.tar.xz
 

Index: linux-6.18.2/scripts/Makefile
===================================================================
--- linux-6.18.2.orig/scripts/Makefile
+++ linux-6.18.2/scripts/Makefile
@@ -6,6 +6,7 @@
 hostprogs-always-$(CONFIG_KALLSYMS)			+= kallsyms
 hostprogs-always-$(BUILD_C_RECORDMCOUNT)		+= recordmcount
 hostprogs-always-$(CONFIG_BUILDTIME_TABLE_SORT)		+= sorttable
+hostprogs-always-$(CONFIG_BUILD_BIN2C)			+= bin2c
 hostprogs-always-$(CONFIG_ASN1)				+= asn1_compiler
 hostprogs-always-$(CONFIG_MODULE_SIG_FORMAT)		+= sign-file
 hostprogs-always-$(CONFIG_SYSTEM_EXTRA_CERTIFICATE)	+= insert-sys-cert
